docs/www: update docusaurus to v3

misterioss opened the pull request:
Closes #5217

## 🎯 Changes
This PR upgrades docusaurus from v2 to v3

## ✅ Checklist
- [X] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

juliusmarminge commented:
Can you fix linting 🙏

KATT commented on www/docusaurus.config.js:
```diff
@@ -22,13 +20,12 @@ const poweredByVercel = `
   url: 'https://trpc.io',
   baseUrl: '/',
-  onBrokenLinks: 'throw',
+  onBrokenLinks: 'ignore',
```
```suggestion
  onBrokenLinks: 'throw',
```

KATT commented on www/docusaurus.config.js:
```diff
@@ -22,13 +20,12 @@ const poweredByVercel = `
   </div>
 `.trim();
 
-/** @type {import('@docusaurus/types').Config} */
```
Can we get types back in?

KATT commented on www/docusaurus.config.js:
```diff
@@ -1,8 +1,6 @@
-
-const { parseEnv } = require('./src/utils/env');
-const { generateTypedocDocusaurusPlugins } = require('./docusaurus.typedoc.js');
+// import type { Config } from '@docusaurus/types';
```
Please get the `// @ts-check` back in
```suggestion
// @ts-check
```

KATT commented on www/docusaurus.preferredTheme.js:
```diff
@@ -14,6 +16,46 @@ if (ExecutionEnvironment.canUseDOM) {
+      makeElementVisible(LIGHT_THEME_SELECTOR);
+      makeElementInvisible(DARK_THEME_SELECTOR);
+    }
+  }
```
Could you please explain how come this is needed?

misterioss replied:
Yes, because the shiki-twoslash support is not as good as for the previous version, so we need to handle theme changing in this way

KATT commented on www/docusaurus.typedoc.js:
```diff
@@ -1,11 +1,6 @@
 
-const fs = require('fs');
-const path = require('path');
-/**
- * @param {string[]} directories
- */
```
Why?

KATT commented on www/docusaurus.typedoc.js:
```diff
@@ -1,11 +1,6 @@
-/* eslint-disable @typescript-eslint/no-var-requires */
-// @ts-check
```
Retain please

misterioss replied:
returned back

KATT commented on www/mdx-to-jsx.js:
How come this is needed? Could it be type-checked with a `// @ts-check`?

misterioss replied:
Yes, there is no good support of Shiki-two slash...
Added // @ts-check

KATT reviewed (commented):
Thanks so much for this!

Please make the branch editable

KATT commented:
Another question, @misterioss, are you able to do ...

```sh
# in root
p install
p build
# switch to `www` and start dev server
cd www && p dev
```

.. on this branch?

It might be some pre-existing issues that got surfaced by this (since our typedoc is kinda whack even now)

misterioss commented:
> Thanks so much for this!
> 
> Please make the branch editable 🙏

Strange, I don's see the checkbox for it...

juliusmarminge commented:
Our custom theme handled diff languague type which the default didnt handle before. Has that been fixed? I saw you reverted to the default one. Deployment fails so cant verify 🤨

KATT reviewed (commented):
Hey

There's still a bunch of issues with typedoc - do you think they are fixable?

We should probably do 

```sh
pnpm add typedoc@latest typedoc-plugin-markdown@latest docusaurus-plugin-typedoc@latest --save-dev
```

in `www/`

KATT commented on www/docusaurus.config.js:
```diff
@@ -25,7 +25,7 @@ const config = {
   url: 'https://trpc.io',
   baseUrl: '/',
-  onBrokenLinks: 'ignore',
+  onBrokenLinks: 'throw',
```
I'm guessing you disable this due to typedoc causing it to throw?

misterioss replied:
Returned back the previous code.

misterioss commented:
@KATT please review again. I've returned ts checks and all possible things. I also merged the latest `next` branch

What was done:
* Upgraded Docusaurus and its dependencies from 2.4.1 to 3.4.0
* Updated Typedoc and its packages from 0.24.7 to 0.26.2
* Updated md/mdx docs files to support v3.0
* Encountered an issue with new Docusaurus and shiki-twoslash support. Implemented temporary workarounds to restore code highlighting as closely as possible to the previous version.

⠀Known issues:
* No copy button for code highlight blocks (because of shiki-twoslash support)
* Sidebar for auto-generated docs (Typedoc issue configuration) order is not as it was before
* Typedoc warnings for @link tags (links need to be updated to the correct format as per https://typedoc.org/tags/link/)

KATT commented:
ohhh, I can see it builds now 👀 👀 🚀 ❤️ 

cc @olehlavryk @MaksRomanchuk, please tick that box so I can
